lib/operations.ts

```diff
@@ -0,0 +1,206 @@
+ * - `stdout (String data)`: When an operation prints to stdout.
+ * - `stderr (String data)`: When an operation prints to stderr.
+ * - `burn (String state)`: When the `burn` operation emits progress state.
+ * - `error (Error error)`: When an error happens.
+ * - `end`: When all the operations are completed successfully.
+ *
+ * @param {String} image - path to image
+ * @param {Object[]} operations - array of operations
+ * @param {Object} options - configuration options
+ *
+ * @returns {EventEmitter}
+ *
+ * @example
+ * const execution = operations.execute(
+ * 	'foo/bar.img',
+ * 	[
+ * 		{
+ * 			command: 'copy',
+ * 			from: {
+ * 				partition: {
+ * 					primary: 1
+ * 				},
+ * 				path: '/bitstreams/parallella_e16_headless_gpiose_7010.bit.bin'
+ * 			},
+ * 			to: {
+ * 				partition: {
+ * 					primary: 1
+ * 				},
+ * 				path: '/parallella.bit.bin'
+ * 			},
+ * 			when: {
+ * 				coprocessorCore: '16',
+ * 				processorType: 'Z7010'
+ * 			}
+ * 		},
+ * 		{
+ * 			command: 'copy',
+ * 			from: {
+ * 				partition: {
+ * 					primary: 1
+ * 				},
+ * 				path: '/bistreams/parallella_e16_headless_gpiose_7020.bit.bin'
+ * 			},
+ * 			to: {
+ * 				partition: {
+ * 					primary: 1
+ * 				},
+ * 				path: '/parallella.bit.bin'
+ * 			},
+ * 			when: {
+ * 				coprocessorCore: '16',
+ * 				processorType: 'Z7020'
+ * 			}
+ * 		}
+ * 	],
+ * 	{
+ * 		coprocessorCore: '16',
+ * 		processorType: 'Z7010'
+ * 	}
+ * );
+ *
+ * execution.on('stdout', process.stdout.write);
+ * execution.on('stderr', process.stderr.write);
+ *
+ * execution.on('state', function(state) {
+ * 	console.log(state.operation.command);
+ * 	console.log(state.percentage);
+ * });
+ *
+ * execution.on('error', function(error) {
+ * 	throw error;
+ * });
+ *
+ * execution.on('end', () => console.log('Finished all operations'));
+ */
+export const execute = function (
+	image: string,
+	operations: Operation[],
+	options?: { os?: string },
+): EventEmitter {
+	options ??= {};
+	options.os ??= utils.getOperatingSystem();
+
+	const missingOptions = utils.getMissingOptions(operations, options);
+
+	if (!_.isEmpty(missingOptions)) {
+		throw new Error(`Missing options: ${missingOptions.join(', ')}`);
+	}
+
+	const emitter: EventEmitter & { ended?: boolean } = new EventEmitter();
+
+	void (async function () {
+		operations = utils.filterWhenMatches(operations, options);
+		const promises = _.map(operations, (operation) =>
+			action.run(image, operation, options),
+		);
+
+		// There is an edge case where the event emitter instance
+		// emits the `end` event before the client is able to
+		// register a listener for it.
+		const emitterOn = emitter.on;
+		emitter.on = function (...args) {
+			const [event, callback] = args;
+			if (event === 'end' && emitter.ended) {
+				// Should this return 'emitterOn' to continue the 'this' chain as per the typings?
```

Comment:
You mean the return value typings of `NodeJS.EventEmitter.on()`? Probably so, I wonder if we have something depending on the different return value. Normally I'd expect to execute the callback but return emitterOn.